### PR TITLE
Enum / Flags: generate user defined `derive`

### PR DESCRIPTION
**Pull request by fengalin**
Allow using API mode TOML config such as:

``` toml
[[object]]
name = "Gst.BufferFlags"
status = "generate"
    [[object.derive]]
    name = "Serialize, Deserialize"
    cfg_condition = "feature = \"ser_de\""

[[object]]
name = "Gst.Format"
status = "generate"
    [[object.derive]]
    name = "Serialize, Deserialize"
    cfg_condition = "feature = \"ser_de\""
```

to generate:

``` rust
 bitflags! {
     #[cfg_attr(feature = "ser_de", derive(Serialize, Deserialize))]
     pub struct BufferFlags: u32 {
         const LIVE = 16;
         [...]
     }
 }
```

and

``` rust
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
 #[cfg_attr(feature = "ser_de", derive(Serialize, Deserialize))]
 pub enum Format {
     Undefined,
     [...]
 }
```

### CODE REVIEW

**Comment by EPashkin** [comment]
@fengalin Code looks good, but problem that in normal struct cfg applied to it, not to next derive.
You sure that it works with bitflags?

**Comment by EPashkin** [comment]
Seems not works: https://play.rust-lang.org/?gist=169b98323d26e8c6412214710d25ee8e&version=stable&mode=debug

Very bad.

**Comment by EPashkin** [comment]
try remove `not`

**Comment by fengalin** [comment]
Yes, you're right! Let me take a look at this...

**Comment by EPashkin** [comment]
It seems possible: 
https://stackoverflow.com/questions/42551113/is-it-possible-to-conditionally-derive-with-features

File: src/config/gobjects.rs [hunk 1]
```diff
@@ -199,6 +203,18 @@ fn parse_object(toml_object: &Value, concurrency: library::Concurrency) -> GObje
     };
     let members = Members::parse(toml_object.lookup("member"), &name);
     let properties = Properties::parse(toml_object.lookup("property"), &name);
+    let derives = {
+        let mut v = Vec::new();
+        if let Some(configs) = toml_object.lookup("derive").and_then(|val| val.as_array()) {
```
Comment (EPashkin) [inline src/config/gobjects.rs]: Maybe this better be as `Derives::parse(`, same as previous?

**Comment by fengalin** [comment]
> It seems possible

Thanks. Implementing it now.

File: src/config/gobjects.rs (hunk 1 above)
Comment (fengalin) [reply src/config/gobjects.rs]: Yes indeed

File: src/config/derives.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+use super::ident::Ident;
+use super::parsable::Parse;
+use toml::Value;
+use super::error::TomlHelper;
+
+#[derive(Clone, Debug)]
+pub struct Derive {
+    pub ident: Ident,
```
Comment (EPashkin) [inline src/config/derives.rs]: Better `pub name: String` instead, but then possible you need add manual parse.

**Comment by EPashkin** [comment]
Working playground for watchers: https://play.rust-lang.org/?gist=0b9aff53230ae4fe6d2248ba3d4d578e&version=stable&mode=debug

File: src/config/derives.rs (hunk 2 above)
Comment (EPashkin) [reply src/config/derives.rs]: `impl<T: Parse + AsRef<Ident>> Parsable for Vec<T> {` just don't need `AsRef` so it IMHO will be applied automatically.

**Comment by fengalin** [comment]
Thanks for the review @EPashkin! It should be better now.

File: src/codegen/flags.rs [hunk 3]
```diff
@@ -74,6 +74,17 @@ fn generate_flags(env: &Env, w: &mut Write, flags: &Bitfield, config: &GObject)
...
+        let derive_str = match &derive.cfg_condition {
+            Some(condition) => format!(
+                "#[cfg_attr(any({}, feature = \"dox\"), derive({}))]",
+                condition,
+                derive.name
+            ),
+            None => format!("#[derive({})]", derive.name),
+        };
+        try!(writeln!(w, "    {}", derive_str));
+    }
```
Comment (EPashkin) [inline src/codegen/flags.rs]: This is duplicate, so please extract it to `codegen\general.rs` with usage `tabs(indent)` instead prefix spaces

**Comment by EPashkin** [comment]
Also better check doc generation, maybe `feature="dox"` will cause error without "ser_de" in real project

**Comment by EPashkin** [comment]
.. and IMHO `[[object.derive]]name = "Serialize, Deserialize" will works too, so readme can be updated to reflect this

**Comment by fengalin** [comment]
> Also better check doc generation, maybe feature="dox" will cause error without "ser_de" in real project

I couldn't find a way to detect doc generation without `feature="dox"` and for some reasons, doc is never generated for `Serialize` and `Deserialize` when used as `#[derive()]`, even without a condition. So I decided to keep only the condition and remove `feature="dox"`. I hope this is acceptable.

File: src/codegen/general.rs [hunk 4]
```diff
@@ -353,6 +354,25 @@ pub fn cfg_condition_string(
     }
 }
 
+pub fn derives(
```
Comment (EPashkin) [inline src/codegen/general.rs]: Please add `#[cfg_attr(feature = "cargo-clippy", allow(ptr_arg))]`

**Comment by EPashkin** [comment]
@fengalin Thanks, only small nit remain.
Yes, without 'dox' is acceptable, as we can write in `cfg_condition` if needed.

Just in case: your old real code compiles with `dox` but without `ser-de` ?
I worried that it just don't compiles as `serde*` crates not included but `#[derive(Serialise)]` is.

File: src/codegen/general.rs (hunk 4 above)
Comment (fengalin) [reply src/codegen/general.rs]: > I fixed the clippy warning.

**Comment by fengalin** [comment]
> your old real code compiles with dox but without ser-de ?

Sorry, my comment was not clear. The code compiles and behaves as expected. It's just the documentation for the `Serialize` and `Deserialize` `Trait` implementations for `enum`s and `struct`s that use `#[derive(Serialize, Deserialize)]` that is not generated.

**Comment by EPashkin** [comment]
Strange.
Thank again, PR is great.

**Pull request merged**